lib/server.js

```diff
@@ -0,0 +1,131 @@
+var Docker = require('dockerode');
+var debug = require('debug')('docker-exec-websocket-server:lib:server');
+var through = require('through');
+var fs = require('fs');
+var ws = require('ws');
+var slugid = require('slugid');
+var _ = require('lodash');
+var path = require('path');
+var assert = require('assert');
+
+var dockerSocket = '/var/run/docker.sock';
+var stats = fs.statSync(dockerSocket);
+if (!stats.isSocket()) {
+  throw new Error('Are you sure the docker is running?');
+}
+var docker = new Docker({socketPath: dockerSocket});
+
+var dbgstream;
+if (process.argv[2] === '--dump') {
+  dbgstream = fs.createWriteStream(path.join(__dirname, '/dump.log'));
+}
+
+// Makes a new buffer with code number (ideally 0x00-0xFF)
+// appended to the start of the given buffer
+function muxBuffer (code, buffer) {
+  return Buffer.concat([new Buffer(String.fromCharCode(code)), buffer]);
+}
+
+/** Makes an instance of docker exec running command, then links the provided
+    websocket to the stream provided by docker exec
+
+    @param tty: boolean representing whether or not terminal emulation is active
+  */
+function execWithSocket (container, socket, command, tty) {
+  var options = {
+    AttachStdout: true,
+    AttachStderr: true,
+    AttachStdin: true,
+    Tty: tty,
+    Detach: false,
+    Cmd: command,
+  };
+  var attachoptions = {
+    stdin: true,
+    stdout: true,
+    stderr: true,
+    stream: true,
+  };
+  return container.exec(options, (err, exec) => {
+    if (err) { //err proof
+      throw err;
+    }
+    exec.start(attachoptions, (err, execStream) => {
+      if (err) { //err proof
+        throw err;
+      }
+      var strout = through((data) => {
+        if (dbgstream) {
+          dbgstream.write('OUT: ' + data + '\n-\n');
+        }
+        socket.send(muxBuffer(1, data), {binary: true, mask: true});
+      });
+      var strerr = through((data) => {
+        if (dbgstream) {
+          dbgstream.write('ERR: ' + data + '\n-\n');
+        }
+        socket.send(muxBuffer(2, data), {binary: true, mask: true});
+      });
+      exec.modem.demuxStream(execStream, strout, strerr);
+
+      socket.on('message', (message) => {
+        if (dbgstream) {
+          dbgstream.write('IN: ' + message + '\n-\n');
+        }
+        execStream.write(message);
```

Comment:
Just fyi, maybe read the docs on [`stream.write`](https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback) for now this is fine. Let's not complicated before we have issues.

But if `execStream.write(message);` returns `false` we might want to tell the sender to not send us anymore before the process has consumed it. Otherwise it'll be buffered up, and we'll potentially run out of memory. I fully expect people to be streaming movies over this socket! :) Okay, well, maybe just big source clones then... who knows..

We probably also want to use the callback that `socket.send` takes [see docs](https://github.com/websockets/ws/blob/master/doc/ws.md#websocketsenddata-options-callback) to make sure that we don't send too many messages..  Ie. if we do we should stop reading from `execStream` and just let the `docker exec` process block when it tries to write to stdout.

Note, this is not the first hack kinda thing, for ssh we don't need flow control, for streaming in/out files or video streams :) we probably will need some form of this in both directions. But it's complicated because we might have to buffer some things to make it fast.